index.js

```diff
@@ -6,38 +6,41 @@ class CleanTerminalPlugin {
 
     this.message = message;
     this.onlyInWatchMode = onlyInWatchMode;
+    this.doClearConsole = false;
   }
 
   apply(compiler) {
-    // Backwards compatible version of compiler.hooks
-    if (!compiler.hooks) {
-      compiler.plugin('emit', (_, done) => {
-        if (this.shouldClearConsole(compiler)) {
-          this.clearConsole();
-        }
-
-        done();
-      });
-
-      return;
+    // Depending on the API available, we either use compiler hooks
+    // or compiler events (compiler events are supported for backwards compatibility)
+    if (compiler.hooks) {
+      this.useCompilerHooks(compiler);
+    } else {
+      this.useCompilerEvents(compiler);
     }
-
-    compiler.hooks.afterCompile.tap('CleanTerminalPlugin', () => {
-      if (this.shouldClearConsole(compiler)) {
-        this.clearConsole();
-      }
-    });
   }
 
   shouldClearConsole(compiler) {
     if (this.onlyInWatchMode) {
-      return compiler.watchMode;
+      return !!compiler.watchMode;
     }
 
-    return (
-      process.env.NODE_ENV !== 'production' &&
-      process.env.options.mode !== 'production'
-    );
+    const isNodeEnvProduction = process.env.NODE_ENV == 'production';
+    const isOptionsModeProduction = process.env.mode && process.env.options.mode == 'production';
```

Comment:
I'm not familiar with `process.env.mode`, but is `process.env.options.mode` only available when `process.env.mode` is `true`?

Reply:
Quite honestly, I never put any though into the two checks. I'm not sure about where they come from. I suspect the env values could come from flags while the options.mode is what's in the config file.

Reply:
Ah right, I can't find any documentation on if, so I guess we should leave it in then. What do you think?

Reply:
If I'm not mistaken @andykais added that part. Maybe he can explain the internals here?

Reply:
I did. Its another mistake of mine I am afraid. That line needs to be changed from `process.env.options.mode` to `compiler.options.mode`. I cannot find documentation on it anywhere, but `compiler.options` is essentially the normalized webpack.config.js object (e.g. its defaults are filled in). So `compiler.options.mode` is how you access the `webpack --mode=` flag, or one defined in the config.

Here is it being used in the [copy-webpack-plugin](https://github.com/webpack-contrib/copy-webpack-plugin/blob/master/src/index.js#L27)

Reply:
Changed as per your feedback. We're now checking if production is either set in the env flags or in the compiler options mode. Running like a kitten :)

Reply:
Ok great, looks good!
Thanks for all the help and input. Nice work!

Going to merge and publish. Let me know if there's anything more to be done, cheers.